### PR TITLE
Include only definitions used by the respective template

### PR DESCRIPTION
**Pull request by ignatiusreza**
Review #316 first

Right now the schema for workflow event and step template includes definitions which it doesn't make use. e.g. definition for button in workflow event; definition for "flow" in step template; etc.. This is extraneous, and will lead to unnecessary large "Your template doesn't match this schema error" during validation..

### CODE REVIEW

File: src/compile/utils.ts [hunk 1]
```diff
@@ -1,17 +1,18 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
 import fs from 'fs';
 import path from 'path';
 import yaml from 'yaml';
 
-import { head, isPlainObject, last, startCase } from 'lodash';
+import { isPlainObject, last, startCase } from 'lodash';
 
+// e.g. '#/definitions/flow/loop' => 'definitions.flow.loop'
```
Comment (kgrhartlage) [inline src/compile/utils.ts:8]:
```suggestion

```

File: src/compile.ts [hunk 2]
```diff
@@ -1,19 +1,57 @@
...
+
+    const pathname = getPropertyPath(value);
+
+    if (has(schema, pathname)) return;
+
+    missingDefinitions.push(pathname);
+  });
+
+  return missingDefinitions;
+};
```
Comment (kgrhartlage) [inline src/compile.ts:28]:
```suggestion
/**
 * Retrieves the paths of referenced definitions within the Workflow Template
 * or Workflow Step Template schema. 
 */
const getReferencedDefinitionsPaths = (schema: Schema) => {
  const references: Array<string> = [];

  forEachDeep(schema, (key, value) => {
    if (key !== '$ref') return;

    const pathname = getPropertyPath(value);

    if (has(schema, pathname)) return;

    references.push(pathname);
  });

  return references;
};
```

File: src/compile.ts [hunk 3]
```diff
@@ -1,19 +1,57 @@
...
+
+const resolveDefinitions = (schema: Schema) => {
+  const missingDefinitions = getMissingDefinitionsSchema(schema);
+
+  missingDefinitions.forEach((definition: string) => {
+    set(schema, definition, get(definitions, definition));
+  });
+
+  return missingDefinitions.length;
+};
```
Comment (kgrhartlage) [inline src/compile.ts:38]:
```suggestion
const resolveDefinitions = (schema: Schema) => {
  const references = getReferencedDefinitionsPaths(schema);

  references.forEach((path: string) => {
    set(schema, path, get(definitions, path));
  });

  return references.length > 0;
};
```

**Review by kgrhartlage — CHANGES_REQUESTED** [review]
only renaming suggestions to add more clarity, apply at your discretion

File: src/compile.ts [hunk 4]
```diff
@@ -1,19 +1,57 @@
...
+
+const assignDefinitions = (rootSchema: Schema) => {
```
Comment (kgrhartlage) [inline src/compile.ts:40]:
```suggestion
const assignDraft07Definitions = (rootSchema: Schema) => {
```

File: src/compile/utils.ts (hunk 1 above)
Comment (ignatiusreza) [reply src/compile/utils.ts:8]: why should it be deleted?

Comment (kgrhartlage) [reply src/compile/utils.ts:8]: my bad, I mistakenly took this for a duplicate

**Review by kgrhartlage — APPROVED** [review]

**Pull request merged**